Split build & packaging

chouquette opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
-->
### What does this PR do?

This PR splits our existing builds in 2 stages:
- A build stage, which is common to all agent version X architecture, which produces an unpackaged intermediary artifact
- A packaging stage, which produces the expected .deb/.rpm

This is described in more details in [this RFC](https://docs.google.com/document/d/1LksLF6magX4rIf_yDkjW9DDB4oKl8XS8HyNQEyRMkeY/edit?pli=1#heading=h.ixldy9ty2xy5)

the build jobs before:
![before](https://github.com/DataDog/datadog-agent/assets/54693/0ac03840-cdba-4af6-82f9-e0b26c049fd8)

and after the split:
![after](https://github.com/DataDog/datadog-agent/assets/54693/9677153d-988e-4e00-9a41-9d381af5526e)

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

Building is long, packaging is cheap, we can save a lot of computing by not building almost identical artifacts for the agent on debian/rhel/suse

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

This is probably hard to review, please feel free to ask for some changes to be extracted in their own PRs.
This is likely to come with minor regressions, which I'd much rather fix iteratively once this is merged instead of adding more changes on this PR.

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

CelianR commented on release.json:
```diff
@@ -8,7 +8,7 @@
         "INTEGRATIONS_CORE_VERSION": "master",
         "OMNIBUS_SOFTWARE_VERSION": "8664357efd9187e1c16284e3d9811ea25d152186",
-        "OMNIBUS_RUBY_VERSION": "78573b961ef3400787a7069526d91795c1a34238",
+        "OMNIBUS_RUBY_VERSION": "chouquette/tmp_split_build_packaging",
```
Did you forget to remove your test ?

chouquette replied:
No, currently omnibus doesn't support some of the things we need, so I need to keep this for the builds to succeed until the changes have been merged there (https://github.com/DataDog/omnibus-ruby/pull/200 & https://github.com/DataDog/omnibus-ruby/pull/201) but indeed these changes will need to be removed eventually

CelianR commented on release.json:
```diff
@@ -27,7 +27,7 @@
         "INTEGRATIONS_CORE_VERSION": "master",
         "OMNIBUS_SOFTWARE_VERSION": "8664357efd9187e1c16284e3d9811ea25d152186",
-        "OMNIBUS_RUBY_VERSION": "78573b961ef3400787a7069526d91795c1a34238",
+        "OMNIBUS_RUBY_VERSION": "chouquette/tmp_split_build_packaging",
```
Did you forget to remove your test ?

KevinFairise2 commented on .gitlab/package_build/deb.yml:
```diff
@@ -1,71 +1,9 @@
 ---
```
If I understand correctly, we might be able to rename the files, since `package_build` should no longer contain things related to `deb/rpm` but only `arm/amd` differences.
Or this one can probably be changed to `heroku.yml` ?

chouquette committed:
move the init script in theiw own software definition

This dill simplify generating what's needed during the packaging stage
instead of generating everything and removing files during the packaging

chouquette committed:
package-artifacts: handle multiple archives

This will happen when we package binaries and debug symbols

chouquette committed:
reorder pkg_metrics and packaging stage

KSerrania commented on omnibus/config/projects/agent.rb:
```diff
@@ -97,6 +97,20 @@
   end
 end
 
+if heroku_target?
+  force_build_and_packaging = true
+  generate_distro_package = true
+else
+  force_build_and_packaging = false
+  if ENV["OMNIBUS_PACKAGE_ARTIFACT_DIR"]
+    dependency "package-artifact"
+    generate_distro_package = true
+    disable_healthcheck true
+  else
+    generate_distro_package = false
+  end
+end
```
I find this hard to read. We have two flags (`a` = we do the build, `b` = we do the packaging), represented by two booleans `force_build_and_packaging = a and b`, `generate_distro_package = b`. I'd have an easier time with one boolean for `a`, and one for `b`.

I think the section would be easier to read and maintain like this, WDYT?

```
build = false
package = false
if ENV["OMNIBUS_PACKAGE_ARTIFACT_DIR"]
  package = true
else
  build = true
end
# Special case for Heroku: we build and package in the same job
if heroku_target?
  build = true
  package = true
end

[...]

package :deb do
  skip_packager !package
  [...]
end

[...]

if build
  dependency 'datadog-agent'
  [...]
  dependency 'datadog-cf-finalize'
end
if package && !build
  disable_healthcheck true
  dependency "package-artifact"
end
if package
  dependency "init-scripts-agent"
end
```

chouquette replied:
that's much clearer indeed, done locally, I'll update the PR shortly

KSerrania commented on omnibus/config/software/datadog-iot-agent.rb:
```diff
@@ -55,33 +55,27 @@
-          mode: 0644,
-          vars: { install_dir: install_dir, etc_dir: etc_dir }
-    end
+    erb source: "upstart_debian.conf.erb",
+        dest: "/etc/init/datadog-agent.conf",
+        mode: 0644,
+        vars: { install_dir: install_dir, etc_dir: etc_dir }
+    # Ship a different upstart job definition on RHEL to accommodate the old
+    # version of upstart (0.6.5) that RHEL 6 provides.
+    erb source: "upstart_redhat.conf.erb",
+        dest: "/etc/init/datadog-agent.conf",
+        mode: 0644,
+        vars: { install_dir: install_dir, etc_dir: etc_dir }
```
Doesn't that mean the redhat init script is always going to be used, as it will overwrite the debian one? I think this should follow the Agent example, and be placed in a separate `init-scripts-iot-agent` definition run at packaging time?

chouquette replied:
indeed, nice catch!

KSerrania reviewed (commented):
Made a first pass, I'll re-review later

chouquette commented on .gitlab/package_build/linux.yml:
```diff
@@ -10,37 +14,30 @@
     - cp $CI_PROJECT_DIR/minimized-btfs.tar.xz /tmp/system-probe/minimized-btfs.tar.xz
     - chmod 0744 /tmp/system-probe/clang-bpf /tmp/system-probe/llc-bpf
-    - inv -e omnibus.build --release-version "$RELEASE_VERSION" --major-version "$AGENT_MAJOR_VERSION" --python-runtimes "$PYTHON_RUNTIMES" --base-dir $OMNIBUS_BASE_DIR  ${USE_S3_CACHING} --skip-deps --go-mod-cache="$GOPATH/pkg/mod" --system-probe-bin=/tmp/system-probe
+    - inv -e omnibus.build --release-version "$RELEASE_VERSION" --major-version "$AGENT_MAJOR_VERSION" --python-runtimes "$PYTHON_RUNTIMES" --base-dir $OMNIBUS_BASE_DIR  ${USE_S3_CACHING} --skip-deps --go-mod-cache="$GOPATH/pkg/mod" --system-probe-bin=/tmp/system-probe --target-project ${DD_TARGET_PROJECT} ${DD_OMNIBUS_ARGS}
```
Indeed, this is a leftover from a previous implementation. Removed

chouquette commented on .gitlab/packaging/deb.yml:
```diff
@@ -1,29 +1,97 @@
+    DD_PKG_ARCH: "x86_64"
+    DD_PROJECT: "agent"
+    RELEASE_VERSION: $RELEASE_VERSION_6
+    AGENT_MAJOR_VERSION: 6
+    PYTHON_RUNTIMES: "2,3"
```
You are most likely correct. This is a leftover from a rather desperate attempt which I didn't cleanup

chouquette committed:
split datadog-agent build and packaging

chouquette committed:
split iot agent build and packaging

chouquette committed:
split dogstatsd build and packaging

chouquette committed:
add a special case for heroku for now

chouquette committed:
fixups for installer

chouquette committed:
tmp omnibus branch

chouquette committed:
generate iot agent init scripts during packaging

chouquette committed:
simplify build/package logic

chouquette commented:
Closing in favor of a PR per product to simplify reviews
